Optimize zfs diff

jwpoduska opened the pull request:
### Motivation and Context
zfs diff was found to be extremely slow to the point of being almost unusable. These changes often improve performance 10 - 100x.

The current implementation of zfs diff uses zfs_obj_to_stats(), which in turn uses zfs_obj_to_path() to get the path to the dnode. The path is then built using zap_value_search() to expand the parents. The problem is zap_value_search() doesn't do a lookup, it does a linear search through the entire directory object, which leads to n^2 behavior.

### Description
This changeset adds the ioctl ZFS_IOC_DUMP_ZAP, which is used to enumerate directories and cache them in userland. The zfs diff stream has been modified to always return a stat object for every id being used since the libzfs_diff code would always do this anyway. A new flag was added to zfs_obj_to_stats() to prevent looking up the path the old way.

These changes are intended to produce identical results to the current diff and thus do not fix any of the problems with the current version (e.g., any of the issues with hard links).

As an extreme example, I ran a large diff on a VM with the following results:
```
root@ub41:~# time zfs diff dattoArray/bfyData/154371@snap0 @snap1 | wc -l
193063
real    212m49.366s
user    0m2.017s
sys     212m44.075s
```
and then with the changes:
```
root@ub41:~# time zfs diff dattoArray/bfyData/154371@snap0 @snap1 | wc -l
193063
real    0m18.139s
user    0m2.008s
sys     0m24.112s
```

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->


### How Has This Been Tested?
The test suite, which includes zfs diff tests was successfully run. In addition, many long running diffs were created to compare both speed and correctness against the current version.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

lundman commented:
Somewhat adjacent to this PR, is there some way to speed up the zap_value_search() implementation we have? OsX has to look up NAME from vget(id), we initially did it with zap_value_search() but it was too slow, we had to change it to cache names in znodes instead. Having a fast name look up in place of zap_value_search would help both issues.

jwpoduska commented:
@lundman it would be nice to have an optimized zap_value_search(), There are issues of course, like id -> path is 1 to many (b/c of links). It also has issues with orphaning, e.g., hard link to an object and then deleting the directory can leave the PARENT entry pointing to a removed objid. This fix is really targeted on improving the performance of zfs diff with its existing warts.

ahrens commented on module/zfs/zfs_ioctl.c:
```diff
@@ -6039,6 +6060,107 @@ zfs_ioc_diff(zfs_cmd_t *zc)
+ * count followed by zap_pair_record_t's to the file descriptor
+ */
+static int
+zfs_ioc_dump_zap(zfs_cmd_t *zc)
```
New ioctls should generally use the new, nvlist-based method.  Could we do that here?

jwpoduska replied:
Makes sense.

ahrens commented:
> zfs diff was found to be extremely slow to the point of being almost unusable. These changes often improve performance 10 - 100x.

My understanding is that we would expect these changes to improve performance only when there are "large" directories.  With more typical size directories, `zfs diff` should not be extremely slow to begin with.  How many directory entries are needed to experience the problem and get an improvement from this PR?

jwpoduska commented:
@ahrens Thanks for looking...

There are a lot of factors including the number of changes that occurred (each one requires a path lookup), the depth of directories (each level requires a lookup) and the size of the zap for each directory. Fragmentation plays a role as well.

On a new pool, the current version of zfs diff can run very fast, but on a mature pool even small directories can take a while. For example, I reran the test from above, but limited the changes to a single directory with 14 files in it. I snapped, deleted the 14 files and timed the result:

```
root@ub41:~# time zfs diff dattoArray/xxxData/123456@snap0 @snap1 | wc -l
15
real    0m12.525s
user    0m0.009s
sys     0m12.405s
```
vs
```
root@ub41:~# time zfs diff dattoArray/xxxData/123456@snap0 @snap1 | wc -l
15
real    0m0.745s
user    0m0.377s
sys     0m0.794s
```

jwpoduska commented on include/sys/zfs_stat.h:
```diff
@@ -44,11 +45,21 @@ typedef struct zfs_stat {
 	uint64_t	zs_mode;
 	uint64_t	zs_links;
 	uint64_t	zs_ctime[2];
+	uint64_t	zs_parent;
+	uint64_t	zs_flags;
```
I'll looks like this would mainly impact zc_zoneid, which is only used by jail/unjail, which only seems to be in freebsd. I could return the stat data (including zs_parent & zs_flags) in an nvlist and keep the size of zfs_cmt_t the same.

jwpoduska commented on include/sys/zfs_ioctl.h:
```diff
@@ -364,20 +365,32 @@ typedef struct dmu_replay_record {
 typedef struct dmu_diff_record {
-	uint64_t ddr_type;
-	uint64_t ddr_first;
-	uint64_t ddr_last;
+	diff_type_t	ddr_type;
+	uint64_t	ddr_obj;
+	int		ddr_err[2];
+	zfs_stat_t	ddr_zs[2];	/* space for 1 or two stats as needed */
 } dmu_diff_record_t;
```
I will look at versioning the stream to prevent this.

jwpoduska commented:
@ahrens I moved the ioctl's to use the nvlist versions, including the diff ioctl, which has the effect of versioning diff as well. So, a mismatched userland/kernel diff will fail, but will not blow up beyond that. I also reverted zfs_stat_t to prevent zfs_cmd_t from changing size.

ColinIanKing commented:
is this change going to be merged at some point?

behlendorf commented:
@ColinIanKing I expect some version of this will be merged after a few rounds of feedback.

ColinIanKing commented:
Apologies for asking again; is this change going to be merged at some point?

terem42 commented:
I've resolved contlicts and created patch for this PR, its available [here](https://github.com/terem42/zfs-debian/blob/master/patches/10391_optimize_zfs_diff.patch). The patch already included into experimental build Debian package using latest master branch, passed ZFS test suite, you can try it out: https://github.com/terem42/zfs-debian
Instruction on Debian repo usage are inside readme.md

jumbi77 commented:
The results looks promising. @behlendorf @ahrens are there any open changes to do? I talked to @jwpoduska and they are using this at datto for a while now... Would be great to get faster zfs diff.

Anyway much thanks!

rincebrain commented:
This would be nice to see - I just added the desire to get something like this to the Hackathon project ideas not having seen this PR (or if I did, having forgot).

If @jwpoduska has no {time, interest} in rebasing this and resolving any outstanding compat issues, I may grab @terem42's rebase and do an updated PR...

jumbi77 commented:
@behlendorf FYI: Considering the number of open PRs i guess this PR can be closed in favor of mentioned #12837 which its the successor of this PR, right?

riedel commented:
Is there a reason why this PR is not followed up upon? I did not realize how slow ` zfs diff `was until I tried it on a slow hard disk raidz pool. I actually aborted it after it ran for more than a day. Calling `zfs diff` unusable in many cases seems no understatement.

rincebrain commented:
If you want to pick up #12837, go for it, I just gave up because it got ignored.

stuartthebruce commented:
FWIW, I just tripped over the current **zfs diff** performance problem again on a system running version 2.3.4. It would be awesome if someone could push this over the finish line.
